substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/parallel/ParallelGCImpl.java

```diff
@@ -0,0 +1,258 @@
+
+    private ChunkBuffer buffer;
+    private volatile boolean inParallelPhase;
+
+    @Fold
+    public static ParallelGCImpl singleton() {
+        return (ParallelGCImpl) ImageSingletons.lookup(ParallelGC.class);
+    }
+
+    public static boolean isInParallelPhase() {
+        return singleton().inParallelPhase;
+    }
+
+    public static void waitForIdle() {
+        singleton().waitForIdleImpl();
+    }
+
+    public static AlignedHeapChunk.AlignedHeader getThreadLocalScannedChunk() {
+        return scannedChunkTL.get();
+    }
+
+    public static AlignedHeapChunk.AlignedHeader getThreadLocalChunk() {
+        return allocChunkTL.get();
+    }
+
+    public static void setThreadLocalChunk(AlignedHeapChunk.AlignedHeader chunk) {
+        allocChunkTL.set(chunk);
+    }
+
+    public void push(Pointer ptr) {
+        buffer.push(ptr);
+    }
+
+    @Override
+    public void startWorkerThreadsImpl() {
+        int hubOffset = ConfigurationValues.getObjectLayout().getHubOffset();
+        VMError.guarantee(hubOffset == 0, "hub offset must be 0");
+
+        // Allocate buffer large enough to store maximum possible number of heap chunks
+        long maxHeapSize = GCImpl.getPolicy().getMaximumHeapSize().rawValue();
+        long alignedChunkSize = HeapParameters.getAlignedHeapChunkSize().rawValue();
+        long unalignedChunkSize = HeapParameters.getLargeArrayThreshold().rawValue();
+        long maxChunks = maxHeapSize / Math.min(alignedChunkSize, unalignedChunkSize) + 1;
+        buffer = new ChunkBuffer((int) maxChunks, ConfigurationValues.getObjectLayout().getReferenceSize());
+
+        workerCount = getWorkerCount();
+        allWorkersBusy = ~(-1 << workerCount) & (-1 << 1);
+        // We reuse the gc thread for worker #0
+        for (int i = 1; i < workerCount; i++) {
+            startWorkerThread(i);
+        }
+    }
+
+    private int getWorkerCount() {
+        int setting = ParallelGCOptions.ParallelGCWorkers.getValue();
+        int workers = setting > 0 ? setting : getDefaultWorkerCount();
+        workers = Math.min(workers, MAX_WORKERS_COUNT);
+        verboseGCLog().string("[Number of ParallelGC workers: ").unsigned(workers).string("]").newline();
+        return workers;
+    }
+
+    private int getDefaultWorkerCount() {
+        // Adapted from Hotspot, see WorkerPolicy::nof_parallel_worker_threads()
+        int cpus = Jvm.JVM_ActiveProcessorCount();
+        return cpus <= 8 ? cpus : 8 + (cpus - 8) * 5 / 8;
+    }
+
+    private void startWorkerThread(int n) {
+        Thread t = new Thread(() -> {
+                VMThreads.SafepointBehavior.markThreadAsCrashed();
+                debugLog().string("WW start ").unsigned(n).newline();
+
+                while (true) {
+                    try {
+                        do {
+                            debugLog().string("WW block ").unsigned(n).newline();
+                            mutex.lock();
```

Comment:
I thought I'd leave that out, since any exception occurring in between would crash the app anyway. Does it make sense to attempt recovery, given that the heap is most likely corrupt?

Reply:
You are right, an exception in that code would cause a crash anyways, so there is no need to attempt recovery. We have the convention to use the try/finally block anyways as it highlights the critical section and makes the code a bit more uniform.